homeassistant/components/nest/climate_sdm.py

```diff
@@ -0,0 +1,331 @@
+"""Support for Google Nest SDM climate devices."""
+
+from typing import Optional
+
+from google_nest_sdm.device import Device
+from google_nest_sdm.device_traits import FanTrait, TemperatureTrait
+from google_nest_sdm.thermostat_traits import (
+    ThermostatEcoTrait,
+    ThermostatHvacTrait,
+    ThermostatModeTrait,
+    ThermostatTemperatureSetpointTrait,
+)
+
+from homeassistant.components.climate import ClimateEntity
+from homeassistant.components.climate.const import (
+    ATTR_TARGET_TEMP_HIGH,
+    ATTR_TARGET_TEMP_LOW,
+    CURRENT_HVAC_COOL,
+    CURRENT_HVAC_HEAT,
+    CURRENT_HVAC_OFF,
+    FAN_OFF,
+    FAN_ON,
+    HVAC_MODE_AUTO,
+    HVAC_MODE_COOL,
+    HVAC_MODE_HEAT,
+    HVAC_MODE_HEAT_COOL,
+    HVAC_MODE_OFF,
+    PRESET_ECO,
+    PRESET_NONE,
+    SUPPORT_FAN_MODE,
+    SUPPORT_PRESET_MODE,
+    SUPPORT_TARGET_TEMPERATURE,
+    SUPPORT_TARGET_TEMPERATURE_RANGE,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_TEMPERATURE, TEMP_CELSIUS
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.typing import HomeAssistantType
+
+from .const import DOMAIN, SIGNAL_NEST_UPDATE
+from .device_info import DeviceInfo
+
+# Mapping for sdm.devices.traits.ThermostatMode mode field
+THERMOSTAT_MODE_MAP = {
+    "OFF": HVAC_MODE_OFF,
+    "HEAT": HVAC_MODE_HEAT,
+    "COOL": HVAC_MODE_COOL,
+    "HEATCOOL": HVAC_MODE_HEAT_COOL,
+}
+THERMOSTAT_INV_MODE_MAP = {v: k for k, v in THERMOSTAT_MODE_MAP.items()}
+
+# Mode for sdm.devices.traits.ThermostatEco
+THERMOSTAT_ECO_MODE = "MANUAL_ECO"
+
+# Mapping for sdm.devices.traits.ThermostatHvac status field
+THERMOSTAT_HVAC_STATUS_MAP = {
+    "OFF": CURRENT_HVAC_OFF,
+    "HEATING": CURRENT_HVAC_HEAT,
+    "COOLING": CURRENT_HVAC_COOL,
+}
+
+# Mapping to determine the trait that supports the target temperatures
+# based on the current HVAC mode
+THERMOSTAT_SETPOINT_TRAIT_MAP = {
+    HVAC_MODE_COOL: ThermostatTemperatureSetpointTrait.NAME,
+    HVAC_MODE_HEAT: ThermostatTemperatureSetpointTrait.NAME,
+    HVAC_MODE_HEAT_COOL: ThermostatTemperatureSetpointTrait.NAME,
+    HVAC_MODE_AUTO: ThermostatEcoTrait.NAME,
+}
+THERMOSTAT_RANGE_MODES = [HVAC_MODE_HEAT_COOL, HVAC_MODE_AUTO]
+
+PRESET_MODE_MAP = {
+    "MANUAL_ECO": PRESET_ECO,
+    "OFF": PRESET_NONE,
+}
+PRESET_INV_MODE_MAP = {v: k for k, v in PRESET_MODE_MAP.items()}
+
+FAN_MODE_MAP = {
+    "ON": FAN_ON,
+    "OFF": FAN_OFF,
+}
+FAN_INV_MODE_MAP = {v: k for k, v in FAN_MODE_MAP.items()}
+
+
+async def async_setup_sdm_entry(
+    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up the client entities."""
+
+    subscriber = hass.data[DOMAIN][entry.entry_id]
+    device_manager = await subscriber.async_get_device_manager()
+
+    # Fetch initial data so we have data when entities subscribe.
```

Comment:
Is this a to do item?

Reply:
Really this is happening under the covers with the device manager above, so I just removed this comment to reduce confusion.  I think this was copied over from when I built this on poll mode weeks ago.